pkg/controller/sparkapplication/sparkapp_metrics.go

```diff
@@ -0,0 +1,233 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sparkapplication
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+	"k8s.io/spark-on-k8s-operator/pkg/apis/sparkoperator.k8s.io/v1alpha1"
+	"k8s.io/spark-on-k8s-operator/pkg/util"
+)
+
+type SparkAppMetrics struct {
+	Labels               []string
+	prefix               string
+	SparkAppSubmitCount  *prometheus.CounterVec
+	SparkAppSuccessCount *prometheus.CounterVec
+	SparkAppFailureCount *prometheus.CounterVec
+	SparkAppRunningCount *util.PositiveGauge
+
+	SparkAppSuccessExecutionTime *prometheus.SummaryVec
+	SparkAppFailureExecutionTime *prometheus.SummaryVec
+
+	SparkAppExecutorRunningCount *util.PositiveGauge
+	SparkAppExecutorFailureCount *prometheus.CounterVec
+	SparkAppExecutorSuccessCount *prometheus.CounterVec
+}
+
+func NewSparkAppMetrics(prefix string, labels []string) *SparkAppMetrics {
+
+	for i, label := range labels {
+		labels[i] = util.CreateValidMetricNameLabel("", label)
+	}
+
+	sparkAppSubmitCount := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_submit_count"),
+			Help: "Spark App Submits via the Operator",
+		},
+		labels,
+	)
+	sparkAppSuccessCount := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_success_count"),
+			Help: "Spark App Success Count via the Operator",
+		},
+		labels,
+	)
+	sparkAppFailureCount := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_failure_count"),
+			Help: "Spark App Failure Count via the Operator",
+		},
+		labels,
+	)
+	sparkAppSuccessExecutionTime := prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_success_execution_time_microseconds"),
+			Help: "Spark App Runtime via the Operator",
+		},
+		labels,
+	)
+	sparkAppFailureExecutionTime := prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_failure_execution_time_microseconds"),
+			Help: "Spark App Runtime via the Operator",
+		},
+		labels,
+	)
+	sparkAppExecutorFailureCount := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_executor_failure_count"),
+			Help: "Spark App Failed Executor Count via the Operator",
+		},
+		labels,
+	)
+	sparkAppExecutorSuccessCount := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_executor_success_count"),
+			Help: "Spark App Successful Executor Count via the Operator",
+		},
+		labels,
+	)
+	sparkAppRunningCount := util.NewPositiveGauge(util.CreateValidMetricNameLabel(prefix, "spark_app_running_count"),
```

Comment:
Why are these two gauges, whereas the other app-related metrics are `CounterVec`?

Reply:
Counters can only be incremented while Gauges can be both incremented/decremented hence for running count, we need Gauges